File: ChickenKiller/src/main/java/botwithus/movement/Movement.java
package botwithus.movement;

import botwithus.navigation.api.NavPath;
import botwithus.navigation.api.State;
import net.botwithus.rs3.world.Coordinate;

public class Movement {
    
    private NavPath navPath;

    public void traverse(NavPath path) {
        if (navPath != null)
            return;
        navPath = path;
        println("Traversing navPath: from: " + path.getStart() + ", to: " + path.getDestination());
    }

    public boolean processNavPath() {
        if (navPath == null)
            return false;
        State state = navPath.state();
        if (state != State.CONTINUE) {
            navPath = null;
            println("Nav path finished with state: " + state);
            return true;
        }
        navPath.process();
        return true;
    }

    public void resetNavPath() {
        if (navPath == null)
            return;
        navPath = null;
        println("Nav path reset.");
    }

    public void navigateToTile(Coordinate toTile, boolean enableSurge, boolean disableTeleports, boolean enableDive) {
        int flags = 0;
        if (enableSurge)
            flags |= NavPath.ENABLE_SURGE;
        if (disableTeleports)
            flags |= NavPath.DISABLE_TELEPORTS;
        if (!enableDive)
            flags |= NavPath.DISABLE_DIVE;
        NavPath path = NavPath.resolve(toTile, flags);
        traverse(path);
    }

    public boolean hasActiveNavPath() {
        return navPath != null;
    }

    private void println(String message) {
        System.out.println(message);
    }
}

File: DooglePicker/src/main/kotlin/botwithus/Movement.kt
package botwithus

import botwithus.navigation.api.NavPath
import botwithus.navigation.api.State
import net.botwithus.rs3.world.Coordinate

class Movement {
    private var navPath: NavPath? = null

    fun traverse(path: NavPath) {
        navPath ?: run {
            navPath = path
            println("Traversing navPath: from: ${path.getStart()}, to: ${path.getDestination()}")
        }
    }

    fun processNavPath(): Boolean {
        val currentPath = navPath ?: return false
        val state = currentPath.state()
        if (state != State.CONTINUE) {
            navPath = null
            println("Nav path finished with state: $state")
            return true
        }
        currentPath.process()
        return true
    }

    fun resetNavPath() {
        navPath = null
        println("Nav path reset.")
    }

    fun navigateToTile(
        toTile: Coordinate?, 
        enableSurge: Boolean, 
        disableTeleports: Boolean, 
        enableDive: Boolean
    ) {
        val flags = buildFlags(enableSurge, disableTeleports, enableDive)
        NavPath.resolve(toTile, flags).let(::traverse)
    }

    private fun buildFlags(enableSurge: Boolean, disableTeleports: Boolean, enableDive: Boolean): Int {
        var flags = 0
        if (enableSurge) flags = flags or NavPath.ENABLE_SURGE
        if (disableTeleports) flags = flags or NavPath.DISABLE_TELEPORTS
        if (!enableDive) flags = flags or NavPath.DISABLE_DIVE
        return flags
    }

    fun hasActiveNavPath(): Boolean = navPath != null

    private fun println(message: String?) = kotlin.io.println(message)
}